edgedb.go

```diff
@@ -154,11 +251,28 @@ func (c *baseConn) Query(
 		args:    args,
 	}
 
-	err = c.granularFlow(ctx, val, q)
+	r, err := c.acquireReader(ctx)
 	if err != nil {
 		return err
 	}
 
+	if e := c.setDeadline(ctx); e != nil {
+		return e
+	}
+
+	err = c.granularFlow(r, val, q)
+	if soc.IsPermenentNetErr(err) {
+		_ = c.conn.Close()
+		c.conn = nil
+	} else {
+		if e := c.setDeadline(context.Background()); e != nil {
+			_ = c.conn.Close()
+			c.conn = nil
+			return wrapAll(err, e)
+		}
+		c.releaseReader(r)
+	}
```

Comment:
Can you create a helper function for this snippet that's repeated in multiple places?

Reply:
I moved all of that into `releaseReader`